projects/sdk-wells/src/lib/Well.ts

```diff
@@ -0,0 +1,449 @@
+import { Well as WellContract } from "src/constants/generated";
+
+import { Auger } from "./Auger";
+import { Pump } from "./Pump";
+import { loadToken, setReadOnly, validateAddress, validateAmount, validateToken } from "./utils";
+import { WellFunction } from "./WellFunction";
+import { WellsSDK } from "./WellsSDK";
+
+export type WellDetails = {
+  tokens: ERC20Token[];
+  wellFunction: WellFunction;
+  pumps: Pump[];
+  auger: Auger;
+};
+
+export type CallStruct = {
+  target: string;
+  data: string;
+};
+
+export type TxOverrides = Overrides & { from?: string };
+
+export class Well {
+  public sdk: WellsSDK;
+  public address: string;
+  public contract: WellContract;
+
+  private name: string | undefined = undefined;
+  private lpToken: ERC20Token | undefined = undefined;
+  private tokens: ERC20Token[] | undefined = undefined;
+  private wellFunction: WellFunction | undefined = undefined;
+  private pumps: Pump[] | undefined = undefined;
+  private auger: Auger | undefined = undefined;
+
+  constructor(sdk: WellsSDK, address: string) {
+    if (!address) {
+      throw new Error("Address must be provided");
+    }
+    setReadOnly(this, "address", address, true);
+    setReadOnly(this, "sdk", sdk, false);
+    setReadOnly(this, "contract", Well__factory.connect(address, sdk.providerOrSigner), false);
+  }
+
+  /**
+   * Retrieves all the well data and stores it in the Well object. Future
+   * getXXX() calls will retrive the local data instead of making an RPC call.
+   * Effectively caches data for:
+   * - getName()
+   * - getWell()
+   * - getLPToken()
+   * - getTokens()
+   * - getWellFunction()
+   * - getPumps()
+   * - getAuger()
+   * - getReserves()
+   * */
+  async loadWell() {
+    await Promise.allSettled([this.getName(), this.getLPToken(), this.getWell(), this.getReserves()]);
+  }
+
+  /**
+   * Get this Well's name
+   */
+  async getName(): Promise<string> {
+    if (!this.name) {
+      setReadOnly(this, "name", await this.contract.name());
+    }
+
+    return this.name!;
+  }
+
+  /**
+   * Get this Well's LP Token
+   */
+  async getLPToken(): Promise<ERC20Token> {
+    if (!this.lpToken) {
+      const token = new ERC20Token(this.sdk.chainId, this.address, undefined, undefined, undefined, this.sdk.providerOrSigner);
+      await token.loadFromChain();
+      setReadOnly(this, "lpToken", token, true);
+    }
+
+    return this.lpToken!;
+  }
+
+  /**
+   * Get the tradeable tokens paired in this Well
+   */
+  async getTokens(): Promise<ERC20Token[]> {
+    if (!this.tokens) {
+      await this.getWell();
+    }
+
+    return this.tokens!;
+  }
+
+  /**
+   * Returns the Well function of this well.
+   * **Well functions** define a relationship between the reserves of the
+   * tokens in the Well and the number of LP tokens.
+   *
+   */
+  async getWellFunction(): Promise<WellFunction> {
+    if (!this.wellFunction) {
+      await this.getWell();
+    }
+
+    return this.wellFunction!;
```

Comment:
I'd expect this to call the well function contract and get the name (presumably e.g. "Constant Product" or whatever).